proxy/src/main/java/net/md_5/bungee/BungeeBossBar.java

```diff
@@ -0,0 +1,276 @@
+package net.md_5.bungee;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.UUID;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import net.md_5.bungee.api.boss.BossBarColor;
+import net.md_5.bungee.api.boss.BossBarDivision;
+import net.md_5.bungee.api.boss.BossBarFlag;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.chat.ComponentSerializer;
+import net.md_5.bungee.protocol.DefinedPacket;
+import net.md_5.bungee.protocol.ProtocolConstants;
+import net.md_5.bungee.protocol.packet.BossBar;
+
+@ToString
+@EqualsAndHashCode
+public class BungeeBossBar implements net.md_5.bungee.api.boss.BossBar
+{
+
+    @Getter
+    private BaseComponent[] title;
+    @Getter
+    private BossBarColor color;
+    @Getter
+    private BossBarDivision division;
+    @Getter
+    private float health;
+    @Getter
+    private boolean visible;
+
+    private List<ProxiedPlayer> players;
+    private EnumSet<BossBarFlag> flags;
+
+    @ToString.Exclude
+    private UUID uuid;
+
+    public BungeeBossBar(BaseComponent[] title, BossBarColor color, BossBarDivision division, float health)
+    {
+        this.title = title;
+        this.color = color;
+        this.division = division;
+        this.health = health;
+        this.visible = true;
+        this.players = new ArrayList<>();
+        this.flags = EnumSet.noneOf( BossBarFlag.class );
+        this.uuid = UUID.randomUUID();
+    }
+
+    @Override
+    public void addPlayer(ProxiedPlayer player)
+    {
+        Preconditions.checkNotNull( player, "player" );
+        if ( !players.contains( player ) )
+        {
+            players.add( player );
+        }
+        if ( player.isConnected() && visible )
+        {
+            player.unsafe().sendPacket( addPacket() );
+        }
+    }
+
+    @Override
+    public void addPlayers(Iterable<ProxiedPlayer> players)
+    {
+        Preconditions.checkNotNull( players, "players" );
+        for ( ProxiedPlayer player : players )
+        {
+            addPlayer( player );
+        }
+    }
+
+    @Override
+    public void removePlayer(ProxiedPlayer player)
+    {
+        Preconditions.checkNotNull( player, "player" );
+        players.remove( player );
+        if ( player.isConnected() && visible )
+        {
+            player.unsafe().sendPacket( removePacket() );
+        }
+    }
+
+    @Override
+    public void removePlayers(Iterable<ProxiedPlayer> players)
+    {
+        Preconditions.checkNotNull( players, "players" );
+        for ( ProxiedPlayer player : players )
+        {
+            removePlayer( player );
+        }
+    }
+
+    @Override
+    public void removeAllPlayers()
+    {
+        removePlayers( players );
```

Comment:
Should clone the list to avoid CME, also the line below becomes redundant. May I also suggest renaming to clearPlayers?

Reply:
> Should clone the list to avoid CME

Sure, but why it is necessary? If the developer actually calls this when a player is added, its his own fault, not ours.

> also the line below becomes redundant

It doesn't as removePlayers accepts a Iterable which cannot and isn't cleared in the method.

> May I also suggest renaming to clearPlayers?

We don't clear the players, we remove them. It is unlogical to rename it like that

Reply:
You're iterating from the same list while removing from the same list at the same time

Reply:
You're iterating from the same list in the method and then when it calls removePlayer it will be removing from the same list consequently

Reply:
Oh wait omg im stupid

Reply:
Ill mark this as resolved